pkg/controller/backup/backup_controller.go

```diff
@@ -0,0 +1,173 @@
+	zero "github.com/infinispan/infinispan-operator/pkg/controller/zerocapacity"
+	"github.com/infinispan/infinispan-operator/pkg/infinispan/backup"
+	"github.com/infinispan/infinispan-operator/pkg/infinispan/client/http"
+	corev1 "k8s.io/api/core/v1"
+	resource "k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+)
+
+const (
+	ControllerName = "backup-controller"
+	DataMountPath  = "/opt/infinispan/backups"
+)
+
+var ctx = context.Background()
+
+// reconcileBackup reconciles a Backup object
+type reconcileBackup struct {
+	// This client, initialized using mgr.Client() above, is a split client
+	// that reads objects from the cache and writes to the apiserver
+	client.Client
+}
+
+type backupResource struct {
+	instance *v2.Backup
+	client   client.Client
+	scheme   *runtime.Scheme
+}
+
+func Add(mgr manager.Manager) error {
+	return zero.CreateController(ControllerName, &reconcileBackup{mgr.GetClient()}, mgr)
+}
+
+func (r *reconcileBackup) ResourceInstance(key client.ObjectKey, ctrl *zero.Controller) (zero.Resource, error) {
+	instance := &v2.Backup{}
+	if err := ctrl.Get(ctx, key, instance); err != nil {
+		return nil, err
+	}
+
+	instance.Spec.ApplyDefaults()
+	return &backupResource{
+		instance: instance,
+		client:   r.Client,
+		scheme:   ctrl.Scheme,
+	}, nil
+}
+
+func (r *reconcileBackup) Type() runtime.Object {
+	return &v2.Backup{}
+}
+
+func (r *backupResource) AsMeta() metav1.Object {
+	return r.instance
+}
+
+func (r *backupResource) Cluster() string {
+	return r.instance.Spec.Cluster
+}
+
+func (r *backupResource) Phase() zero.Phase {
+	return zero.Phase(string(r.instance.Status.Phase))
+}
+
+func (r *backupResource) UpdatePhase(phase zero.Phase) error {
+	instance := r.instance
+	instance.Status.Phase = v2.BackupPhase(string(phase))
+	return r.updateStatus()
+}
+
+func (r *backupResource) Init() (*zero.Spec, error) {
+
+	// TODO add labels
+	pvc := &corev1.PersistentVolumeClaim{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      r.instance.Name,
+			Namespace: r.instance.Namespace,
+		},
+	}
+
+	volumeSpec := r.instance.Spec.Volume
+	var storage resource.Quantity
+	if volumeSpec.Storage == nil {
+		// TODO calculate based upon number of Pods in cluster
+		// ISPN- Utilise backup size estimate
+		storage = constants.DefaultPVSize
+	} else {
+		storage = resource.MustParse(*volumeSpec.Storage)
+	}
+
+	_, err := controllerutil.CreateOrUpdate(ctx, r.client, pvc, func() error {
+		pvc.Spec = corev1.PersistentVolumeClaimSpec{
+			AccessModes: []corev1.PersistentVolumeAccessMode{
+				corev1.ReadWriteOnce,
+			},
+			Resources: corev1.ResourceRequirements{
+				Requests: corev1.ResourceList{
+					corev1.ResourceStorage: storage,
+				},
+			},
+			StorageClassName: volumeSpec.StorageClassName,
```

Comment:
You should setup storageClassName if it only defined by user. In best approach it would be nice to validate it before as this can't be updated later in case of incorrect defined

Reply:
`BackupVolumeSpec.StorageClassName` is a pointer to a string, if the user doesn't specify anthing then `nil` is passed to the `PersistentVolumeClaimSpec`, which is the equivalent of not setting it. Or am I missing something?

Reply:
I'm not sure, but seems to be a reason of this error on OCP
```
{"level":"error","ts":1602086957.1748934,"logger":"controller-runtime.controller","msg":"Reconciler error","controller":"backup-controller","request":"infinispan-operator/example-infinispan-backup","error":"Unable to create pvc: PersistentVolumeClaim \"example-infinispan-backup\" is invalid: spec: Forbidden: spec is immutable after creation except resources.requests for bound claims\n  core.PersistentVolumeClaimSpec{\n  \t... // 2 identical fields\n  \tResources:        core.ResourceRequirements{Requests: core.ResourceList{\"storage\": {i: resource.int64Amount{value: 1073741824}, s: \"1Gi\", Format: \"BinarySI\"}}},\n  \tVolumeName:       \"\",\n- \tStorageClassName: nil,\n+ \tStorageClassName: &\"gp2\",\n  \tVolumeMode:       &\"Filesystem\",\n  \tDataSource:       nil,\n  }\n","stacktrace":"github.com/go-logr/zapr.(*zapLogger).Error\n\t/home/dvolodin/go/pkg/mod/github.com/go-logr/zapr@v0.1.1/zapr.go:128\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\n\t/home/dvolodin/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.4.0/pkg/internal/controller/controller.go:258\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/home/dvolodin/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.4.0/pkg/internal/controller/controller.go:232\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).worker\n\t/home/dvolodin/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.4.0/pkg/internal/controller/controller.go:211\nk8s.io/apimachinery/pkg/util/wait.JitterUntil.func1\n\t/home/dvolodin/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20191004115801-a2eda9f80ab8/pkg/util/wait/wait.go:152\nk8s.io/apimachinery/pkg/util/wait.JitterUntil\n\t/home/dvolodin/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20191004115801-a2eda9f80ab8/pkg/util/wait/wait.go:153\nk8s.io/apimachinery/pkg/util/wait.Until\n\t/home/dvolodin/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20191004115801-a2eda9f80ab8/pkg/util/wait/wait.go:88"}
```